client/basic-authorship/src/basic_authorship.rs

```diff
@@ -395,4 +402,47 @@ mod tests {
 			storage_changes.transaction_storage_root,
 		);
 	}
+
+	#[test]
+	fn should_not_remove_invalid_transactions_when_skipping() {
+		// given
+		let client = Arc::new(substrate_test_runtime_client::new());
+		let txpool = Arc::new(
+			BasicPool::new(Default::default(), Arc::new(FullChainApi::new(client.clone()))).0
+		);
+
+		futures::executor::block_on(
+			txpool.submit_at(&BlockId::number(0), vec![
+				extrinsic(0),
+				extrinsic(1),
+				Transfer {
+					amount: Default::default(),
+					nonce: 2,
+					from: AccountKeyring::Alice.into(),
+					to: Default::default(),
+				}.into_exhaust_tx(),
+				extrinsic(3),
+				extrinsic(4),
+				extrinsic(5),
+			])
+		).unwrap();
+
+		let mut proposer_factory = ProposerFactory::new(client.clone(), txpool.clone());
+		let mut proposer = proposer_factory.init_with_now(
+			&client.header(&BlockId::number(0)).unwrap().unwrap(),
+			Box::new(move || time::Instant::now()),
+		);
+
+		// when
+		let deadline = time::Duration::from_secs(9);
+		let block = futures::executor::block_on(
+			proposer.propose(Default::default(), Default::default(), deadline, RecordProof::No)
+		).map(|r| r.block).unwrap();
+
+		// then
+		// block should have some extrinsics although we have some more in the pool.
+		assert_eq!(block.extrinsics().len(), 2);
+		assert_eq!(txpool.ready().count(), 6);
```

Comment:
I still would like to see that we build a second block here.

Reply:
Then it means that the transaction cannot always exhaust resources. I guess it's fine, I'll change it so that it exhaust resources if it's not the ONLY/FIRST transaction in the block. Then keeping that variant as `Transfer` extension makes even more sense to me.